--- leveldb_base/table/table_builder.cc	2026-04-24 22:39:53
+++ leveldb/table/table_builder.cc	2026-04-24 13:13:15
@@ -4,12 +4,14 @@
 
 #include "leveldb/table_builder.h"
 
+#include "db/dbformat.h"
 #include <cassert>
 
 #include "leveldb/comparator.h"
 #include "leveldb/env.h"
 #include "leveldb/filter_policy.h"
 #include "leveldb/options.h"
+
 #include "table/block_builder.h"
 #include "table/filter_block.h"
 #include "table/format.h"
@@ -26,6 +28,7 @@
         offset(0),
         data_block(&options),
         index_block(&index_block_options),
+        range_del_block(&options),
         num_entries(0),
         closed(false),
         filter_block(opt.filter_policy == nullptr
@@ -42,6 +45,7 @@
   Status status;
   BlockBuilder data_block;
   BlockBuilder index_block;
+  BlockBuilder range_del_block;
   std::string last_key;
   int64_t num_entries;
   bool closed;  // Either Finish() or Abandon() has been called.
@@ -95,8 +99,20 @@
   Rep* r = rep_;
   assert(!r->closed);
   if (!ok()) return;
-  if (r->num_entries > 0) {
+  if (r->num_entries > 0 && !r->last_key.empty()) {
     assert(r->options.comparator->Compare(key, Slice(r->last_key)) > 0);
+  }
+
+  size_t n = key.size();
+  assert(n >= 8);
+  const uint64_t tag = DecodeFixed64(key.data() + n - 8);
+  ValueType type = static_cast<ValueType>(tag & 0xff);
+
+  if (type == kTypeRangeDeletion) {
+    r->range_del_block.Add(key, value);
+    r->last_key.assign(key.data(), key.size());
+    r->num_entries++;
+    return;
   }
 
   if (r->pending_index_entry) {
@@ -216,8 +232,15 @@
   assert(!r->closed);
   r->closed = true;
 
-  BlockHandle filter_block_handle, metaindex_block_handle, index_block_handle;
+  BlockHandle filter_block_handle, metaindex_block_handle, index_block_handle,
+      range_del_block_handle;
 
+  bool has_range_dels = !r->range_del_block.empty();
+
+  if (ok() && !r->range_del_block.empty()) {
+    WriteBlock(&r->range_del_block, &range_del_block_handle);
+  }
+
   // Write filter block
   if (ok() && r->filter_block != nullptr) {
     WriteRawBlock(r->filter_block->Finish(), kNoCompression,
@@ -236,6 +259,13 @@
       meta_index_block.Add(key, handle_encoding);
     }
 
+    if (has_range_dels) {
+      std::string key = "leveldb.range_deletions";
+      std::string handle_encoding;
+      range_del_block_handle.EncodeTo(&handle_encoding);
+      meta_index_block.Add(key, handle_encoding);
+    }
+
     // TODO(postrelease): Add stats and other meta blocks
     WriteBlock(&meta_index_block, &metaindex_block_handle);
   }
