--- leveldb_base/include/leveldb/db.h	2026-04-24 22:39:53
+++ leveldb/include/leveldb/db.h	2026-04-24 21:26:17
@@ -7,6 +7,9 @@
 
 #include <cstdint>
 #include <cstdio>
+#include <string>
+#include <utility>
+#include <vector>
 
 #include "leveldb/export.h"
 #include "leveldb/iterator.h"
@@ -87,6 +90,13 @@
   virtual Status Get(const ReadOptions& options, const Slice& key,
                      std::string* value) = 0;
 
+  virtual Status Scan(
+      const ReadOptions& options, const Slice& start_key, const Slice& end_key,
+      std::vector<std::pair<std::string, std::string>>* result) = 0;
+
+  virtual Status DeleteRange(const WriteOptions& options,
+                             const Slice& start_key, const Slice& end_key) = 0;
+
   // Return a heap-allocated iterator over the contents of the database.
   // The result of NewIterator() is initially invalid (caller must
   // call one of the Seek methods on the iterator before using it).
@@ -145,6 +155,8 @@
   // Therefore the following call will compact the entire database:
   //    db->CompactRange(nullptr, nullptr);
   virtual void CompactRange(const Slice* begin, const Slice* end) = 0;
+
+  virtual Status ForceFullCompaction() = 0;
 };
 
 // Destroy the contents of the specified database.
