--- leveldb_base/db/memtable.h	2026-04-24 22:39:53
+++ leveldb/db/memtable.h	2026-04-25 20:10:39
@@ -5,11 +5,12 @@
 #ifndef STORAGE_LEVELDB_DB_MEMTABLE_H_
 #define STORAGE_LEVELDB_DB_MEMTABLE_H_
 
-#include <string>
-
 #include "db/dbformat.h"
 #include "db/skiplist.h"
+#include <string>
+
 #include "leveldb/db.h"
+
 #include "util/arena.h"
 
 namespace leveldb {
@@ -62,6 +63,10 @@
   // Else, return false.
   bool Get(const LookupKey& key, std::string* value, Status* s);
 
+  const RangeDeletionList* GetRangeDeletions() const {
+    return &range_deletions_;
+  }
+
  private:
   friend class MemTableIterator;
   friend class MemTableBackwardIterator;
@@ -80,6 +85,7 @@
   int refs_;
   Arena arena_;
   Table table_;
+  RangeDeletionList range_deletions_;
 };
 
 }  // namespace leveldb
