--- leveldb_base/db/memtable.cc	2026-04-24 22:39:53
+++ leveldb/db/memtable.cc	2026-04-26 07:35:22
@@ -3,10 +3,13 @@
 // found in the LICENSE file. See the AUTHORS file for names of contributors.
 
 #include "db/memtable.h"
+
 #include "db/dbformat.h"
+
 #include "leveldb/comparator.h"
 #include "leveldb/env.h"
 #include "leveldb/iterator.h"
+
 #include "util/coding.h"
 
 namespace leveldb {
@@ -97,13 +100,21 @@
   std::memcpy(p, value.data(), val_size);
   assert(p + val_size == buf + encoded_len);
   table_.Insert(buf);
+
+  if (type == kTypeRangeDeletion) {
+    range_deletions_.Add(key, value, s);
+  }
 }
 
 bool MemTable::Get(const LookupKey& key, std::string* value, Status* s) {
   Slice memkey = key.memtable_key();
   Table::Iterator iter(&table_);
   iter.Seek(memkey.data());
-  if (iter.Valid()) {
+
+  const Slice ikey = key.internal_key();
+  SequenceNumber read_seq = DecodeFixed64(ikey.data() + ikey.size() - 8) >> 8;
+
+  while (iter.Valid()) {
     // entry format is:
     //    klength  varint32
     //    userkey  char[klength]
@@ -120,7 +131,15 @@
             Slice(key_ptr, key_length - 8), key.user_key()) == 0) {
       // Correct user key
       const uint64_t tag = DecodeFixed64(key_ptr + key_length - 8);
-      switch (static_cast<ValueType>(tag & 0xff)) {
+      ValueType type = static_cast<ValueType>(tag & 0xff);
+      SequenceNumber found_seq = tag >> 8;
+
+      if (range_deletions_.IsDeleted(key.user_key(), found_seq, read_seq)) {
+        *s = Status::NotFound(Slice());
+        return true;
+      }
+
+      switch (type) {
         case kTypeValue: {
           Slice v = GetLengthPrefixedSlice(key_ptr + key_length);
           value->assign(v.data(), v.size());
@@ -129,9 +148,20 @@
         case kTypeDeletion:
           *s = Status::NotFound(Slice());
           return true;
+        case kTypeRangeDeletion:
+          iter.Next();
+          continue;
       }
+    } else {
+      break;
     }
   }
+
+  if (range_deletions_.IsDeleted(key.user_key(), 0, read_seq)) {
+    *s = Status::NotFound(Slice());
+    return true;
+  }
+
   return false;
 }
 
