--- leveldb_base/db/version_set.h	2026-04-24 22:39:53
+++ leveldb/db/version_set.h	2026-04-25 00:42:40
@@ -15,12 +15,12 @@
 #ifndef STORAGE_LEVELDB_DB_VERSION_SET_H_
 #define STORAGE_LEVELDB_DB_VERSION_SET_H_
 
+#include "db/dbformat.h"
+#include "db/version_edit.h"
 #include <map>
 #include <set>
 #include <vector>
 
-#include "db/dbformat.h"
-#include "db/version_edit.h"
 #include "port/port.h"
 #include "port/thread_annotations.h"
 
@@ -111,12 +111,17 @@
 
   int NumFiles(int level) const { return files_[level].size(); }
 
+  const std::vector<FileMetaData*>& files_at_level(int level) const {
+    return files_[level];
+  }
+
   // Return a human readable string that describes this version's contents.
   std::string DebugString() const;
 
  private:
   friend class Compaction;
   friend class VersionSet;
+  friend class DBImpl;
 
   class LevelFileNumIterator;
 
@@ -390,4 +395,4 @@
 
 }  // namespace leveldb
 
-#endif  // STORAGE_LEVELDB_DB_VERSION_SET_H_
+#endif  // STORAGE_LEVELDB_DB_VERSION_SET_H_
\ No newline at end of file
