--- leveldb_base/db/version_set.cc	2026-04-24 22:39:53
+++ leveldb/db/version_set.cc	2026-04-25 20:10:39
@@ -4,16 +4,17 @@
 
 #include "db/version_set.h"
 
-#include <algorithm>
-#include <cstdio>
-
 #include "db/filename.h"
 #include "db/log_reader.h"
 #include "db/log_writer.h"
 #include "db/memtable.h"
 #include "db/table_cache.h"
+#include <algorithm>
+#include <cstdio>
+
 #include "leveldb/env.h"
 #include "leveldb/table_builder.h"
+
 #include "table/merger.h"
 #include "table/two_level_iterator.h"
 #include "util/coding.h"
@@ -246,12 +247,7 @@
 
 // Callback from TableCache::Get()
 namespace {
-enum SaverState {
-  kNotFound,
-  kFound,
-  kDeleted,
-  kCorrupt,
-};
+enum SaverState { kNotFound, kFound, kDeleted, kCorrupt, kDeletedByRange };
 struct Saver {
   SaverState state;
   const Comparator* ucmp;
@@ -266,7 +262,20 @@
     s->state = kCorrupt;
   } else {
     if (s->ucmp->Compare(parsed_key.user_key, s->user_key) == 0) {
-      s->state = (parsed_key.type == kTypeValue) ? kFound : kDeleted;
+      switch (parsed_key.type) {
+        case kTypeValue:
+          s->state = kFound;
+          break;
+        case kTypeDeletion:
+          s->state = kDeleted;
+          break;
+        case kTypeRangeDeletion:
+          s->state = kDeletedByRange; // Halt search for this key
+          break;
+        default:
+          s->state = kCorrupt;
+          break;
+      }
       if (s->state == kFound) {
         s->value->assign(v.data(), v.size());
       }
@@ -365,6 +374,7 @@
           state->found = true;
           return false;
         case kDeleted:
+        case kDeletedByRange:
           return false;
         case kCorrupt:
           state->s =
